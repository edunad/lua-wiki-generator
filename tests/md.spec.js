#!/usr/bin/env node
'use strict';

const fs = require('fs-extra');
const { resolve } = require('path');

const MDParser = require('../src/md.js');
const LuaParser = require('../src/parser.js');

const METHOD_TEMPLATE_FILE = './__test_templates__/METHOD_TEMPLATE.md';
const EXTENSION_TEMPLATE_FILE = './__test_templates__/EXTENSION_TEMPLATE.md';
const CLASS_TEMPLATE_FILE = './__test_templates__/CLASS_TEMPLATE.md';
const GVAR_TEMPLATE_FILE = './__test_templates__/GVAR_TEMPLATE.md';
const CUSTOM_METHOD_TEMPLATE_FILE = './__test_templates__/CUSTOM_METHOD_TEMPLATE.md';
const BAD_TEMPLATE_FILE = './__test_templates__/BAD_TEMPLATE.md';

const OUTPUT_FOLDER = './readme';

describe('MDMethod', () => {
    beforeAll(async () => {
        this.templates = {
            method: await fs.readFile(resolve(__dirname, METHOD_TEMPLATE_FILE), 'utf8'),
            extension: await fs.readFile(resolve(__dirname, EXTENSION_TEMPLATE_FILE), 'utf8'),
            class: await fs.readFile(resolve(__dirname, CLASS_TEMPLATE_FILE), 'utf8'),
            gvar: await fs.readFile(resolve(__dirname, GVAR_TEMPLATE_FILE), 'utf8'),
            custom_method: await fs.readFile(resolve(__dirname, CUSTOM_METHOD_TEMPLATE_FILE), 'utf8'),

            invalid: await fs.readFile(resolve(__dirname, BAD_TEMPLATE_FILE), 'utf8'),
        };
    });

    beforeEach(() => {
        MDParser.setTextMDParser(null);
        MDParser.setLinkMDParser(null);
        MDParser.setHintMDParser(null);
    });

    it('Successfully generates a class MD file', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.class_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.class, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully generates a method MD file', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.method_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.method, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully generates a extension MD file', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.extension_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.extension, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully generates a gvar MD file', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.gvar_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.gvar, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully handles unknown fields', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.method_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.invalid, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully handles missing fields', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.bad_method_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.method, data.blocks[0])).toMatchSnapshot();
    });

    it('Successfully handles custom parsers', async () => {
        const parser = new LuaParser(resolve(__dirname, `./__test_lua__/ias.method_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        MDParser.setTextMDParser((outputFolder, template, block) => {
            return [true, template.replace('$MY_CUSTOM_METHOD$', block.type)];
        });

        MDParser.setLinkMDParser((type, outputFolder, data) => {
            if (type === '$TITLE_NAME$') {
                return `[TITLE](My title link)`;
            }

            return template;
        });

        MDParser.setHintMDParser((hint) => {
            return `:::\n${hint.type} -> ${hint.message}\n`;
        });

        expect(MDParser.generate(OUTPUT_FOLDER, this.templates.custom_method, data.blocks[0])).toMatchSnapshot();
    });
});
