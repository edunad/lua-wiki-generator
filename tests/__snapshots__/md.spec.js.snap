// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDMethod Successfully generates a class MD file 1`] = `
"# AABB

### <img src="../../.gitbook/assets/base.png" width="32" height="32" /> AABB
AABB's entry class<br/>

-----------------
## Fields

| Type   | Name | Optional |
| ------ | ---- | -------: |
| [Vector](vector/README.md) | pos |   |
| [Vector](vector/README.md) | size |   |
|  | meep |   |
"
`;

exports[`MDMethod Successfully generates a extension MD file 1`] = `
"# ~~math.isNan~~

### <img src="../../.gitbook/assets/client.png" width="32" height="32" /> ~~math.isNan~~

> âš  Deprecated: Do not use this anymore

:::success
OH NOOOOOOO
:::


\`\`\`lua
boolean math.isNan(val)
\`\`\`

Returns true if the number is NaN (not a number)<br/>It also does some other magic<br/>

-----------------
## Parameters

| Type   | Name | Description | Optional |
| ------ | ---- | ----------- | -------: |
| number | val | No description |   |

-----------------
## Returns

| Type   | Description |
| ------ | ----------: |
| boolean | If its NaN |
"
`;

exports[`MDMethod Successfully generates a gvar MD file 1`] = `
"# _G.SERVER

### <img src="../../.gitbook/assets/global.png" width="32" height="32" /> **_G**.SERVER

:::info
For SHARED code, just don't check for SERVER or CLIENT
:::

Returns true if the current code is running on the SERVER<br/>
"
`;

exports[`MDMethod Successfully generates a method MD file 1`] = `
"# console:execute

### <img src="../../.gitbook/assets/shared.png" width="32" height="32" /> console:execute

:::warning
PAGE / FUNCTIONALITY STILL IN CONSTRUCTION
:::
:::danger
OH NOOOOOOO
:::


\`\`\`lua
boolean, string console:execute(args, var)
\`\`\`

Executes a console command programmatically<br/>
\`\`\`lua
--my_command is the command to execute, true is a parameter
console:execute({"my_command", "true"})
\`\`\`


-----------------
## Parameters

| Type   | Name | Description | Optional |
| ------ | ---- | ----------- | -------: |
| string[] | args | No description |   |
| number | var | No description |   |
| string | var | No description |   |
| boolean | var | No description |   |

-----------------
## Returns

| Type   | Description |
| ------ | ----------: |
| boolean | If command was executed successfully |
| string | No description |
"
`;

exports[`MDMethod Successfully handles custom parsers 1`] = `
"## ![shared](.gitbook/assets/shared.png) [TITLE](My title link)

METHOD

:::
warning -> PAGE / FUNCTIONALITY STILL IN CONSTRUCTION
:::
danger -> OH NOOOOOOO


"
`;

exports[`MDMethod Successfully handles missing fields 1`] = `
"# [console](readme//README.md):execute

### <img src="../../.gitbook/assets/unknown.png" width="32" height="32" /> [console](readme//README.md):execute

:::uhhhh
OH NOOOOOOO
:::


\`\`\`lua
void console:execute(args)
\`\`\`

Executes a console command programmatically<br/>

-----------------
## Parameters

| Type   | Name | Description | Optional |
| ------ | ---- | ----------- | -------: |
| string[]" | "args | No description |   |
"
`;

exports[`MDMethod Successfully handles unknown fields 1`] = `
"##![Base](.gitbook/assets/base.png) $TITLES_B$

![$RANDOM$](.gitbook/assets/$RANDOM$.png)
$UNKNOWN_FIELD$
"
`;
